from __future__ import print_function

from argon2.exceptions import VerifyMismatchError
from flask_restful import Resource
from flask import request
from models import Usuario
from db import db
from flask_jwt_extended import jwt_required, create_access_token, get_jwt, get_jwt_identity, JWTManager
from helpers import encryptPassword, checkPassword
from marshmallow.exceptions import ValidationError
from common.error_handling import ObjectNotFound, NotAllowed

jwt = JWTManager()


class SignInResource(Resource):
    def post(self):
        if request.json["contrasena1"] != request.json["contrasena2"]:
            raise NotAllowed("Las contraseñas no coinciden")
        nuevo_usuario = Usuario(username=request.json["username"],
                                contrasena=encryptPassword(request.json["contrasena1"]),
                                email=request.json["email"])
        db.session.add(nuevo_usuario)
        db.session.commit()
        token_de_acceso = create_access_token(identity=nuevo_usuario.id)
        return {"mensaje": "usuario creado exitosamente", "token": token_de_acceso}



class LogInResource(Resource):
    def post(self):
        try:
            usuario = Usuario.query.filter(Usuario.username == request.json["username"]).first()
            if usuario is None:
                return {"mensaje": "El usuario no existe"}, 404
            else:
                checkPassword(usuario.contrasena, request.json["password"])
                token_de_acceso = create_access_token(identity=usuario.id)
                return {"mensaje": "Inicio de sesión exitoso", "token": token_de_acceso}
        except ValidationError as e:
            return {"mensaje": e.messages[0]}, 401
        except VerifyMismatchError as e:
            return {"mensaje": "Incorrect user or password"}, 404
